### PR TITLE
sql: tie time functions to the gateway clock instead of the HLC

### LINKED ISSUE #5767: sql: tie timestamp/date functions to local clock instead of hlc clock
From https://github.com/cockroachlabs/registration/issues/17#issuecomment-203936108

> I think the HLC is doing the right thing for the way we use it internally, but I'm not sure that exposing it (or rather, its wall-time component) as the sql-level timestamp is the right thing. Guaranteeing a non-decreasing clock at the expense of a much higher rate of duplicate values doesn't feel like a very good tradeoff. Maybe we should revisit cockroachdb/cockroach#5037 and instead of changing the basic timestamp functions, we introduce a new function that returns a strictly increasing HLC timestamp.

The concrete proposal is to revert the various time functions (`now()`, `current_timestamp()`, `current_date(), `age(), `transaction_timestamp()`, `statement_timestamp()`, etc) to use the node's local clock and introduce a new function (e.g. `cluster_timestamp()`) which is taken from the hlc clock.

@knz, @bdarnell Thoughts?

### PR DESCRIPTION
**Pull request by knz**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5805)

### CODE REVIEW

**Comment by tschottdorf** [comment]
Haven't reviewed yet, but would it make sense to have `cluster_timestamp` return something that's an actual HLC timestamp (as opposed to splitting the hlc in two components, which mostly just invites running into trouble)? Is there any "somewhat" natural data type that this would map to (of course we could always do a bigint of type `((walltime << 32)+logical)`).

**Comment by knz** [comment]
Ok that's a thought but that would remove the opportunity to do time computations in SQL (intervals etc.) Is this desirable?

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEHdP9GZaduAuADvJVd:33402988)*

**Comment by tschottdorf** [comment]
While you can compute intervals between HLC timestamps, they don't really mean anything (after all, HLC is closer to a logical clock), so I would intuitively view removing support for that as positive.
Or is there some information in HLC intervals that I might be missing?

**Comment by knz** [comment]
In particular the duration  `cluster_timestamp() - transaction_timestamp()` gives you how much time a txn was waiting/busy with retries.

**Comment by tschottdorf** [comment]
That's not really true, and even if it were, the computation `cluster_timestamp() >> 32 - transaction_timestamp()` isn't that hard (assuming that's what we go for).

**Comment by knz** [comment]
(I think it is true since the txn timestamp is populated initially also from the HLC)

But otherwise yes i see your argument of course we can measure this using integers. Let's make this function an int.

**Comment by tschottdorf** [comment]
No, if your HLC is ahead of the physical wall clock then that's all you're measuring, even in the absence of contention. You could similarly argue that it can measure clock offsets.

**Comment by knz** [comment]
Ok so `cluster_timestamp()` is now DECIMAL. The Logical field becomes its fractional part.

**Comment by knz** [comment]
Test for this patch needs fix for #5812

**Comment by bdarnell** [comment]
Exposing the cluster timestamp as a decimal is definitely better than two separate values, but I still don't think it makes sense to do arithmetic on these values. The only operation that's well-defined on HLC timestamps is ordering, so I'd prefer to expose them as (opaque but sortable) 12-byte blobs.

---

Reviewed 3 of 6 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/eval.go, line 817 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KENlRKg16PIjxlDBk3W:-KENlRKhpzTdS9PDJI_t:-563195317) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1a8ca14a11b9ec1198c3b66a50b1e5c296ae12a/sql/parser/eval.go#L817)):*
Why ten billion? `WallTime` is in nanoseconds; shouldn't it be 1 billion?

There should be some tests of the values computed here. Maybe factor out the `roachpb.Timestamp` to decimal conversion into a separate function so you can test conversion on known values. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KENmbrOGJFbIV1JJCvU:2028639540)*

**Comment by knz** [comment]
I beg to disagree on two fronts.
First of all one thing that makes sense is substraction: you can meaningfully order transactions by their timestamp distance to a reference transaction.

Then also remember the whole thing was motivated by the need to be able to test ordering client side. In the case I'm coming from the client doesn't do comparisons/sorts on blobs, only numbers. 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/eval.go, line 817 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KENlRKg16PIjxlDBk3W:-KEVJhbLuer7NrgBLi5P:-1676379529) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1a8ca14a11b9ec1198c3b66a50b1e5c296ae12a/sql/parser/eval.go#L817)):*
:+1: on testing.

The operation performed here is (WallTime + 10^10 + Logical).Scale(10)

Why 10? Logical has 32 bits, so its maximum value needs 10 decimal digits.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEVIyPdIOxVb9vF94Hz:-1202377297)*

**Comment by tschottdorf** [comment]
Ordering is what @bdarnell already points out *is* well-defined, but it has nothing to do with their *distance* (a naive version of which is subtracting the timestamps). So I definitely don't want anyone to do arithmetic on these values thinking it's meaningful. But opaque 12 byte blobs are also not really useful. If my specific use case dictates use of cluster timestamps, it'd be nice to see the walltime portion in cleartext as it can be helpful in debugging (though maybe the same can be achieved if these 12 byte blobs could be string-formatted for human consumption, but that's more code to write and document).

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEWXuPT8IADwc7YeXij:-74572682)*

**Comment by bdarnell** [comment]
You can order two transactions by their distance to a reference transaction as long as that reference transaction is the same everywhere, but how is that any better than just measuring everything relative to the epoch? As soon as you are computing deltas relative to two different baselines the differences become impossible to interpret (unless you're talking about time scales that are much larger than MaxOffset. But in that case you could just use real wall time instead of HLC)

If your client can't compare blobs that sounds like it should be fixed on the client.

I'd argue that pretty-printing HLC timestamps can be misleading for all the reasons we're discussing here, so putting it behind a formatting function might be a good idea. (a decimal timestamp is also not very user-friendly)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/eval.go, line 817 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KENlRKg16PIjxlDBk3W:-KEWokoL8OohYeTjNu-r:-1097716530) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1a8ca14a11b9ec1198c3b66a50b1e5c296ae12a/sql/parser/eval.go#L817)):*
Ah, I misunderstood what was going on here as a second-to-nanosecond conversion, not creating space for the logical bits. This could use a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEWoxusR1g_bXJYI0mF:1501008014)*

**Comment by knz** [comment]
Ok I did not express myself properly. The issue is that the values are large and unyieldy. In order to appear as appropriately-sized, we can take the ts of a reference transaction (doesn't matter when, as long as it's close by in time) then substract it (server-side, as part of the query) from the ts of subsequent transactions, yielding a small value (in absolute value) that can be casted to int/float without loss of precision and which has the same monotonicity property. 

Also I did not consider what Toby said but for troubleshooting it is handy to see numbers as well.

---

Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/parser/eval.go, line 817 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KENlRKg16PIjxlDBk3W:-KEXHzTMqYIe-DJhdw8J:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1a8ca14a11b9ec1198c3b66a50b1e5c296ae12a/sql/parser/eval.go#L817)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEXGecB27Y2YoqgyF3P:864969234)*

**Comment by knz** [comment]
Tests fixed thanks to resolution of #5812.

**Comment by petermattis** [comment]
Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/session.go, line 178 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEf_OLM4hio8CsjVhk1:-KEf_OLM4hio8CsjVhk2:-1151948659) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e558f88cf3bd1e0907a6052056a4444a1d60a7/sql/session.go#L178)):*
`s/throughout each/for the lifetime of a/g`

---

*[sql/parser/eval.go, line 759 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEf_q7tIpKzOVMph1jD:-KEf_q7tIpKzOVMph1jE:-1757811278) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e558f88cf3bd1e0907a6052056a4444a1d60a7/sql/parser/eval.go#L759)):*
`s/.  /. /g` (remove double spaces)
`s/cluster_timestamp/cluster_logical_timestamp/g`
`s/throughout the/for the lifetime of the/g`

---

*[sql/parser/eval.go, line 832 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEfa7hYHTXnjf7nKQip:-KEfa7hYHTXnjf7nKQiq:727245962) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e558f88cf3bd1e0907a6052056a4444a1d60a7/sql/parser/eval.go#L832)):*
You're performing all this work for every transaction. Let's move this to `GetClusterTimestamp()` so that only callers of `cluster_logical_timestamp()` bear the burden. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805)*

**Comment by knz** [comment]
Review status: 1 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/session.go, line 178 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEf_OLM4hio8CsjVhk1:-KEfd-Qx-3Dfvej7OiZn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e558f88cf3bd1e0907a6052056a4444a1d60a7/sql/session.go#L178)):*
Done.

---

*[sql/parser/eval.go, line 759 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEf_q7tIpKzOVMph1jD:-KEfd2KPBKxCFDklErBr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e558f88cf3bd1e0907a6052056a4444a1d60a7/sql/parser/eval.go#L759)):*
Done.

---

*[sql/parser/eval.go, line 832 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEfa7hYHTXnjf7nKQip:-KEfd3m2OU8myP-pK9sY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/94e558f88cf3bd1e0907a6052056a4444a1d60a7/sql/parser/eval.go#L832)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/parser/eval.go, line 832 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEfdZMsrslkXXOoVksh:-KEfdZMsrslkXXOoVksi:1089712111) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8a7fbe2bf55aaab648db0f6b8fa7a6c0004e832/sql/parser/eval.go#L832)):*
`s/10000000000/1e10/g`, so the reader doesn't have to count the zeroes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEfdqg26PknMUOSfCkd:1417088721)*

**Comment by knz** [comment]
Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/parser/eval.go, line 832 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEfdZMsrslkXXOoVksh:-KEfnfQwVdBUr_jH0iTf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8a7fbe2bf55aaab648db0f6b8fa7a6c0004e832/sql/parser/eval.go#L832)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805)*

**Comment by tamird** [comment]
the commit message doesn't follow convention in a number of ways and has some grammatical issues. other than that, :lgtm:

---

Reviewed 1 of 6 files at r1, 1 of 2 files at r3, 2 of 3 files at r4, 1 of 2 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/session.go, line 178 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEgUKPY28xQuN2MYaot:-KEgUKPZB16ecJCpRiZf:1763985537) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5011507d1b755f0b60a285f0736beb9510a302f/sql/session.go#L178)):*
s/Must/must/

---

*[sql/parser/eval.go, line 786 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-KEgVAePWrm9E8YfPdIV:-KEgVAePWrm9E8YfPdIW:746685952) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5011507d1b755f0b60a285f0736beb9510a302f/sql/parser/eval.go#L786)):*
missing period

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5805#-:-KEgUUQWb1DeRQ7oy2Kl:573217591)*

File: sql/executor.go [hunk 1]
```diff
@@ -373,7 +373,9 @@ func (e *Executor) execRequest(ctx context.Context, session *Session, sql string
 			txnState.reset(ctx, e, session)
 			txnState.State = Open
 			txnState.autoRetry = true
-			execOpt.MinInitialTimestamp = e.ctx.Clock.Now()
+			now := e.ctx.Clock.Now()
+			txnState.sqlTimestamp = now
```
Comment (tschottdorf) [inline sql/executor.go]: Sorry for being late to the party, but now you're (again) taking a timestamp off the HLC for the statement timestamp, which means it's going to behave just like the actual transaction timestamp (including not changing for up to `MaxOffset`). It was my understanding that we wanted a timestamp of the local *physical* clock, i.e. `e.ctx.Clock.Physical{Now,Time}()` (which would not be a HLC timestamp, but only a wall time).

**Comment by knz** [comment]
@tschottdorf yes you are right this is an oversight from my side. I had overlooked that the HLC's Now() stalls on the same wall time if a transaction pushed it forward. I will submit a follow-up PR.

**Pull request merged**